File: src/main/java/com/multicampus/gamesungcoding/a11ymarketserver/common/advice/GlobalExceptionHandler.java
package com.multicampus.gamesungcoding.a11ymarketserver.common.advice;

import com.multicampus.gamesungcoding.a11ymarketserver.common.advice.model.ErrorRespStatus;
import com.multicampus.gamesungcoding.a11ymarketserver.common.advice.model.RestErrorResponse;
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.*;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.auth.dto.LoginErrResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(UserNotFoundException.class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public ResponseEntity<RestErrorResponse> handleUserNotFoundException(UserNotFoundException ex) {
        log.error(ex.getMessage());
        return ResponseEntity
                .status(HttpStatus.NOT_FOUND)
                .body(new RestErrorResponse(
                        ErrorRespStatus.USER_NOT_FOUND,
                        ex.getMessage())
                );
    }

    @ExceptionHandler(DataNotFoundException.class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public ResponseEntity<RestErrorResponse> handleDataNotFoundException(DataNotFoundException ex) {
        log.error(ex.getMessage());
        return ResponseEntity
                .status(HttpStatus.NOT_FOUND)
                .body(new RestErrorResponse(
                        ErrorRespStatus.DATA_NOT_FOUND,
                        ex.getMessage())
                );
    }

    @ExceptionHandler(DataDuplicatedException.class)
    @ResponseStatus(HttpStatus.CONFLICT)
    public ResponseEntity<RestErrorResponse> handleDuplicatedDataException(DataDuplicatedException ex) {
        log.error(ex.getMessage());
        return ResponseEntity
                .status(HttpStatus.CONFLICT)
                .body(new RestErrorResponse(
                        ErrorRespStatus.DUPLICATED_DATA,
                        ex.getMessage())
                );
    }

    @ExceptionHandler(InvalidRequestException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public ResponseEntity<RestErrorResponse> handleInvalidRequestException(InvalidRequestException ex) {
        log.error(ex.getMessage());
        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(new RestErrorResponse(
                        ErrorRespStatus.INVALID_REQUEST,
                        ex.getMessage())
                );
    }

    @ExceptionHandler(LoginFailedException.class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public ResponseEntity<LoginErrResponse> handleLoginFailedException(LoginFailedException ex) {
        log.error(ex.getMessage());
        return ResponseEntity
                .status(HttpStatus.NOT_FOUND)
                .body(new LoginErrResponse(
                        ex.getMessage())
                );
    }
}

File: src/main/kotlin/com/multicampus/gamesungcoding/a11ymarketserver/common/advice/GlobalExceptionHandler.kt
package com.multicampus.gamesungcoding.a11ymarketserver.common.advice

import com.multicampus.gamesungcoding.a11ymarketserver.common.advice.model.ErrorRespStatus
import com.multicampus.gamesungcoding.a11ymarketserver.common.advice.model.RestErrorResponse
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.*
import com.multicampus.gamesungcoding.a11ymarketserver.feature.auth.dto.LoginErrResponse
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {
    private val log = LoggerFactory.getLogger(this::class.java)

    private fun buildError(status: HttpStatus, errorStatus: ErrorRespStatus, message: String) =
        ResponseEntity.status(status).body(RestErrorResponse(status.value(), errorStatus, message))

    @ExceptionHandler(UserNotFoundException::class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    fun handleUserNotFoundException(ex: UserNotFoundException) =
        buildError(
            HttpStatus.NOT_FOUND,
            ErrorRespStatus.USER_NOT_FOUND,
            ex.message ?: "기본 에러 메시지"
        ).also { log.error(ex.message) }

    @ExceptionHandler(DataNotFoundException::class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    fun handleDataNotFoundException(ex: DataNotFoundException) =
        buildError(
            HttpStatus.NOT_FOUND,
            ErrorRespStatus.DATA_NOT_FOUND,
            ex.message ?: "사용자를 찾을 수 없습니다."
        ).also { log.error(ex.message) }

    @ExceptionHandler(DataDuplicatedException::class)
    @ResponseStatus(HttpStatus.CONFLICT)
    fun handleDuplicatedDataException(ex: DataDuplicatedException) =
        buildError(
            HttpStatus.CONFLICT,
            ErrorRespStatus.DUPLICATED_DATA,
            ex.message ?: "데이터가 이미 존재합니다."
        ).also { log.error(ex.message) }

    @ExceptionHandler(InvalidRequestException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleInvalidRequestException(ex: InvalidRequestException) =
        buildError(
            HttpStatus.BAD_REQUEST,
            ErrorRespStatus.INVALID_REQUEST,
            ex.message ?: "잘못된 요청입니다."
        ).also { log.error(ex.message) }

    @ExceptionHandler(LoginFailedException::class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    fun handleLoginFailedException(ex: LoginFailedException) = ResponseEntity
        .status(HttpStatus.NOT_FOUND)
        .body<LoginErrResponse?>(
            LoginErrResponse(
                ex.message ?: "로그인에 실패했습니다."
            )
        ).also { log.error(ex.message) }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleMethodArgumentNotValidException(ex: MethodArgumentNotValidException) =
        buildError(
            HttpStatus.BAD_REQUEST,
            ErrorRespStatus.INVALID_REQUEST,
            "잘못된 요청입니다. "
        ).also { log.error(ex.message) }
}